### PR TITLE
vmclone: use patchset and remove informers from struct

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
vmclone controller pkg cleanup:
1. Use the patch package to replace the hardcoded patch operations. This improves the readability and make the code more cohesive.
2. Remove the usage of the informers in the vmclone controller.
There is no need to hold informers, narrowing down
the interface can decrease the probability of making
a mistake.

Partially address: https://github.com/kubevirt/kubevirt/issues/11654

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/apimachinery/patch/patch.go [hunk 1]
```diff
@@ -119,6 +119,18 @@ func (p *PatchSet) GeneratePayload() ([]byte, error) {
 	return GeneratePatchPayload(p.patches...)
 }
 
+func (p *PatchSet) GenerateStringPatchOperations() ([]string, error) {
```
Comment (xpivarc) [inline pkg/apimachinery/patch/patch.go:122]: Do you need this? Isn't simple conversion enough?

File: pkg/virt-controller/watch/clone/clone_base.go [hunk 2]
```diff
@@ -70,7 +71,11 @@ func NewVmCloneController(client kubecli.KubevirtClient, vmCloneInformer, snapsh
 		cloneStatusUpdater:      status.NewCloneStatusUpdater(client),
 	}
 
-	_, err := ctrl.vmCloneInformer.AddEventHandler(
+	ctrl.hasSynced = func() bool {
+		return vmCloneInformer.HasSynced() && snapshotInformer.HasSynced() && restoreInformer.HasSynced() && vmInformer.HasSynced()
```
Comment (xpivarc) [inline pkg/virt-controller/watch/clone/clone_base.go]: What about these `napshotContentInformer, pvcInformer`?

**Review by xpivarc — CHANGES_REQUESTED** [review]

File: pkg/virt-controller/watch/clone/clone_base.go (hunk 2 above)
Comment (fossedihelm) [reply pkg/virt-controller/watch/clone/clone_base.go]: see next commit :)

File: pkg/apimachinery/patch/patch.go (hunk 1 above)
Comment (EdDev) [reply pkg/apimachinery/patch/patch.go:122]: IMO it does not fit as an API addition to PatchSet.
If the logic is unique to some specific part of the code, it should be done there.

I would avoid doing it here to assure not to encourage using it in other places.

Comment (fossedihelm) [reply pkg/apimachinery/patch/patch.go:122]: Agreed! That was one of my doubts actually, will move it. Thanks

**Comment by fossedihelm** [comment]
@EdDev @xpivarc PTAL

**Comment by alicefr** [comment]
The code looks good
/lgtm

**Review by EdDev — COMMENTED** [review]
Regarding the `PatchSet` object change, I would prefer if @alicefr can approve/review.

/hold

File: pkg/apimachinery/patch/patch.go [hunk 3]
```diff
@@ -77,6 +77,10 @@ func New(opts ...PatchOption) *PatchSet {
 	return p
 }
 
+func (p *PatchSet) GetPatches() []PatchOperation {
```
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/apimachinery/patch/patch.go:80]: @alicefr already gave lgtm.

**Pull request merged**